lib/charms/prometheus_k8s/v0/prometheus_remote_write.py

```diff
@@ -1,93 +1,701 @@
             for unit in relation.units:
-                # If external-address is provided use that, else use ingress-address
-                if (address := relation.data[unit].get("address")) is None:
+                if unit.app is self._charm.app:
+                    # This is a peer unit
                     continue
-                if (port := relation.data[unit].get("port")) is None:
-                    continue
-                yield f"http://{address}:{port}/api/v1/write"
+
+                if remote_write := relation.data[unit].get("remote_write"):
+                    deserialized_remote_write = json.loads(remote_write)
+                    endpoints.append(
+                        {
+                            "url": deserialized_remote_write["url"],
+                        }
+                    )
+
+        return endpoints
+
+    def _label_alert_topology(self, rule) -> dict:
+        """Insert juju topology labels into an alert rule.
+
+        Args:
+            rule: a dictionary representing a prometheus alert rule.
+
+        Returns:
+            a dictionary representing prometheus alert rule with juju
+            topology labels.
+        """
+        metadata = self._consumer_metadata
+        labels = rule.get("labels", {})
+        labels["juju_model"] = metadata["model"]
+        labels["juju_model_uuid"] = metadata["model_uuid"]
+        labels["juju_application"] = metadata["application"]
+        rule["labels"] = labels
+        return rule
+
+    def _label_alert_expression(self, rule) -> dict:
+        """Insert juju topology filters into a prometheus alert rule.
+
+        Args:
+            rule: a dictionary representing a prometheus alert rule.
+
+        Returns:
+            a dictionary representing a prometheus alert rule that filters based
+            on juju topology.
+        """
+        metadata = self._consumer_metadata
+        topology = 'juju_model="{}", juju_model_uuid="{}", juju_application="{}"'.format(
+            metadata["model"], metadata["model_uuid"], metadata["application"]
+        )
+
+        if expr := rule.get("expr", None):
+            expr = expr.replace("%%juju_topology%%", topology)
+            rule["expr"] = expr
+        else:
+            logger.error("Invalid alert expression in %s", rule.get("alert"))
+
+        return rule
 
     @property
-    def configs(self) -> Generator[dict, None, None]:
-        """A config object ready to be dropped in to a prometheus config file.
+    def _labeled_alert_groups(self) -> list:
+        """Load alert rules from rule files.
+
+        All rules from files for a consumer charm are loaded into a single
+        group. the generated name of this group includes juju topology
+        prefixes.
 
         Returns:
-            A list of remote_write configs.
+            a list of prometheus alert rule groups.
+        """
+        alerts = []
+        for path in Path(self._alert_rules_path).glob("*.rule"):
```

Comment:
Potentially worth changing this. An innocuous `import os.path as path`, if it ever slips in, will lead to really annoying debugging and unpredictable behavior.

It's not C, but `for p in Path(...` is Pythonic enough